hedera-node/src/main/java/com/hedera/services/fees/charging/FeeChargingPolicy.java

```diff
@@ -0,0 +1,134 @@
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+
+/**
+ * Provides the transaction fee-charging policy for the processing
+ * logic. The policy offers four basic entry points:
+ * <ol>
+ *    <li>For a txn whose submitting node seemed to ignore due diligence
+ *    (e.g. submitted a txn with an impermissible valid duration); and, </li>
+ *    <li>For a txn that looks to have been submitted responsibly, but is
+ *    a duplicate of a txn already submitted by a different node; and,</li>
+ *    <li>For a triggered txn; and,</li>
+ *    <li>For a txn that was submitted responsibly, and is believed unique.</li>
+ * </ol>
+ *
+ * @author Michael Tinker
+ */
+public class FeeChargingPolicy {
+	private final NarratedCharging narratedCharging;
+
+	public FeeChargingPolicy(NarratedCharging narratedCharging) {
+		this.narratedCharging = narratedCharging;
+	}
+
+	/**
+	 * Apply the fee charging policy to a txn that was submitted responsibly, and
+	 * believed unique.
+	 *
+	 * @param fees the fee to charge
+	 * @return the outcome of applying the policy
+	 */
+	public ResponseCodeEnum apply(FeeObject fees) {
+		return chargePendingSolvency(fees);
+	}
+
+	/**
+	 * Apply the fee charging policy to a txn that was submitted responsibly, but
+	 * is a duplicate of a txn already submitted by a different node.
+	 *
+	 * @param fees the fee to charge
+	 * @return the outcome of applying the policy
+	 */
+	public ResponseCodeEnum applyForDuplicate(FeeObject fees) {
+		final var feesForDuplicate = new FeeObject(fees.getNodeFee(), fees.getNetworkFee(), 0L);
+
+		return chargePendingSolvency(feesForDuplicate);
+	}
+
+	/**
+	 * Apply the fee charging policy to a txn that was submitted responsibly, but
+	 * is a triggered txn rather than a parent txn requiring node precheck work.
+	 *
+	 * @param fees the fee to charge
+	 * @return the outcome of applying the policy
+	 */
+	public ResponseCodeEnum applyForTriggered(FeeObject fees) {
+		narratedCharging.setFees(fees);
+
+		if (!narratedCharging.isPayerWillingToCoverServiceFee()) {
+			return INSUFFICIENT_TX_FEE;
+		} else if (!narratedCharging.canPayerAffordServiceFee()) {
+			return INSUFFICIENT_PAYER_BALANCE;
+		} else {
+			narratedCharging.chargePayerServiceFee();
+			return OK;
+		}
+	}
+
+	/**
+	 * Apply the fee charging policy to a txn that looks to have been
+	 * submitted without performing basic due diligence.
+	 *
+	 * @param fees the fee to charge
+	 * @return the outcome of applying the policy
+	 */
+	public ResponseCodeEnum applyForIgnoredDueDiligence(FeeObject fees) {
+		narratedCharging.setFees(fees);
+		narratedCharging.chargeSubmittingNodeUpToNetworkFee();
+		return OK;
+	}
+
+	private ResponseCodeEnum chargePendingSolvency(FeeObject fees) {
+		narratedCharging.setFees(fees);
+
+		if (!narratedCharging.isPayerWillingToCoverNetworkFee()) {
+			narratedCharging.chargeSubmittingNodeUpToNetworkFee();
+			return INSUFFICIENT_TX_FEE;
```

Comment:
So the logic here is that if the payer offers a large amount of fee willing to pay, but it actually has almost zero balance, in this cause, the submitting node will be charged and the payer walked away with nothing to lose?

Reply:
The node should check that the payer will be able to afford the network fee before it submit's the txn...if it doesn't check, it is penalized for this lack of "due diligence".

Reply:
I see. If it does and the transaction goes through, the submitting node will get reward as compensation.